### PR TITLE
`error-stack` support `Debug` hooks in `no-std` contexts

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

While exploring #1399 we found the [spin crate](https://crates.io/crates/spin), which implements a `no-std` equivalent for `std::sync` types, like `RwLock` based on spin locks. This PR adds a new feature, `hooks`, which - if enabled - will replace the `RwLock` with the spin variant.

Note: while we expect a majority of calls to be read-only, there's still a chance of priority inversion, this is why if both `std` and `hooks` features are enabled, we will instead use the `std` `RwLock`.

> Must read for spin locks and their application: https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html

I consider using a spinlock okay for `error-stack` hooks, as we expect hooks to only be initialized very, very infrequently, and most operations are reads. A read is simply a `fetch_add` + `fetch_sub` without any spinning.

An alternative approach would be using [`critical-section`](https://crates.io/crates/critical-section) instead, which acts like a `Mutex`, but because we mostly only read, we do not need to hold the hooks exclusively, and `critical-section` is only available on single thread embedded microprocessors.

> **Note:** this increases our test permutation count to 81!

## 🔗 Related links

- https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html

## 🔍 What does this change?

- new feature `hook`

## 📋 TODO

- [x] Documentation
- [x] Changelog

## 📜 Does this require a change to the docs?

We need to mention the new feature in our table and the consequences.

### CODE REVIEW

File: packages/libs/error-stack/Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ anyhow = { version = "1.0.65", default-features = false, optional = true }
 eyre = { version = "0.6", default-features = false, optional = true }
 owo-colors = { version = "3", default-features = false, optional = true, features = ['supports-colors'] }
 serde = { version = "1", default-features = false, optional = true }
+spin = { version = "0.9.4", default-features = false, optional = true, features = ['rwlock', 'once'] }
```
Comment (TimDiekmann) [inline packages/libs/error-stack/Cargo.toml:25]: Does `0.9` also work for our use cases?

**Review by TimDiekmann — COMMENTED** [review]
LGTM, thank you!

Do we need to adjust the Makefile to also test `hooks` on `insta` checks?

One other question, then I'm happy to approve

**Comment by indietyp** [comment]
Thanks! I completely forgot about the snapshot tests, I have adjusted the Makefile and test cfg guard

File: packages/libs/error-stack/Cargo.toml (hunk 1 above)
Comment (indietyp) [reply packages/libs/error-stack/Cargo.toml]: requirement relaxed in [122a119](https://github.com/hashintel/hash/pull/1556/commits/122a119a12021b17f17782443d0422b47826f166)

Comment (TimDiekmann) [reply packages/libs/error-stack/Cargo.toml]: Great, thanks!

**Review by TimDiekmann — APPROVED** [review]
Excellent, thank you, and also for providing the link to the blog post.

**Pull request merged**